pipelines/test/ldd-check.yaml

```diff
@@ -38,51 +38,94 @@ pipeline:
+        [ "$VERBOSE" = "false" ] || echo "$@"
+      }
+
+      failormsg() {
+        local cond="$1" msg="$2"
+        case "$cond" in
+          true|false) :;;
+          *) error "cond=$cond - expected true or false";;
+        esac
+        [ "$cond" = "true" ] && fail "$msg"
+        vmsg "$msg"
+      }
+
+      check_output() {
+        local file="$1" outf="$2" errf="$3"
         missing=$(awk \
-          '$0 ~ /=> not found/ { miss = miss " " $1; }; END { printf("%s\n", miss); }' \
-          "$outf") || error "$f: parsing with awk failed $?";
-        if [ "$verbose" = "true" ]; then
+          '$0 ~ /=> not found/ { miss = miss " " $1; };
+           END { printf("%s\n", miss); }' "$outf") ||
+             error "$f: parsing with awk failed $?";
+        if [ "$VERBOSE" = "true" ]; then
           echo "> $ ldd $f"
           sed 's,^,> ,' "$outf"
         fi
-        [ -z "$missing" ] && { pass "$f"; continue; }
-        fail "$f: missing ${missing# }"
+        [ -z "$missing" ] || fail "$f: missing ${missing# }"
+        pass "$f"
+      }
+
+      check_file() {
+        local f="$1" insist_dyn="$2" rc="0"
+        local outf="$tmpd/ldd.stdout" errf="$tmpd/ldd.sterr"
+        if [ ! -e "$f" ]; then
+          failormsg "$insist_dyn" "$f: did not exist"
+          return 0
+        fi
+        if [ ! -f "$f" ]; then
+          failormsg "$insist_dyn" "$f: is not a file"
+          return 0
+        fi
+
+        ldd "$f" >$outf 2>$errf || rc=$?
+
+        if [ $rc -eq 1 ]; then
+          if grep -q 'not a dynamic executable' "$errf"; then
+            failormsg "$insist_dyn" "$f: not a dynamic exectuable"
+            return 0
+          fi
+          echo "> $ ldd $f"
+          sed 's,^,> ,' "$outf" "$errf"
+          error "$f: ldd exited $rc"
+        elif [ $rc -ne 0 ]; then
+          echo "> $ ldd $f"
+          sed 's,^,> ,' "$outf" "$errf"
+          error "$1: unexpected exit code $rc"
+        fi
+
+        check_output "$f" "$outf" "$errf"
       }
 
       test_files_in() {
+        local f=""
         echo "[ldd-check] Testing binaries in package $pkg"
-        apk info -eq "$pkg" > /dev/null || \
-          { fail "Package $pkg is not installed"; continue; }
+        apk info -eq "$pkg" > /dev/null ||
+          error "Package $pkg is not installed";
         apk info -Lq "$pkg" > "$tmpd/$pkg.list"
         while read f; do
           [ -n "$f" ] || continue
-          f="/$f"
-          [ -f "$f" ] || continue
-          ldd "$f" > /dev/null 2>&1 || continue
-          test_file "$f"
+          check_file "/$f" false
         done < "$tmpd/$pkg.list"
       }
 
       set -- $files
       for f in "$@"; do
-        test_file "$f"
+        check_file "$f" true
       done
       set -- $packages
       for pkg in "$@"; do
         test_files_in "$pkg"
       done
-      echo "tested $((passes+fails)) files with ldd. $passes passes. $fails fails."
+      echo "tested $((passes+fails)) files with ldd." \
```

Comment:
I really wish this  line was prefaced with "[ldd-check]" like some of the other lines so one could easily grep log files for "[ldd-check]".

Reply:
https://github.com/wolfi-dev/os/pull/39042/commits/69f73a83c353e185990e438850681eeaffa657f3